File: _DEMO/CoolHitDetector.cs
using UnityEngine;
using System;
using System.Collections.Generic;

namespace DEMO
{
    /// <summary>
    /// Detects "cool hits" worthy of demo clip extraction.
    /// Evaluates hits based on distance, velocity, prediction accuracy, and other criteria.
    /// </summary>
    public class CoolHitDetector : MonoBehaviour
    {
        [Header("Detection Thresholds")]
        [SerializeField] private float longRangeDistance = 100f;      // Distance threshold for "long range"
        [SerializeField] private float highVelocityThreshold = 50f;  // Velocity threshold for "high speed"
        [SerializeField] private float predictionThreshold = 2f;      // Lead distance for "prediction shot"
        [SerializeField] private int minBouncesForCool = 2;          // Minimum bounces for "bank shot"

        [Header("Clip Settings")]
        [SerializeField] private float clipBeforeHit = 2f;            // Seconds before hit to include
        [SerializeField] private float clipAfterHit = 1f;             // Seconds after hit to include
        [SerializeField] private int maxClipsPerSession = 20;         // Limit number of clips per session

        private DemoRecorder _recorder;
        private List<DemoClip> _detectedClips;
        private int _clipsThisSession;

        public event Action<DemoClip> OnCoolHitDetected;

        public int DetectedClipCount => _detectedClips.Count;
        public DemoClip[] DetectedClips => _detectedClips.ToArray();

        private void Awake()
        {
            _detectedClips = new List<DemoClip>();
            _clipsThisSession = 0;
        }

        private void Start()
        {
            _recorder = FindAnyObjectByType<DemoRecorder>();
            if (_recorder == null)
            {
                Debug.LogWarning("[CoolHitDetector] No DemoRecorder found");
            }
        }

        /// <summary>
        /// Evaluate a projectile hit and determine if it's "cool".
        /// </summary>
        public bool EvaluateHit(ProjectileEvent hitEvent, ProjectileEvent fireEvent, int bounceCount)
        {
            if (hitEvent.eventType != ProjectileEventType.Hit)
                return false;

            CoolHitCriteria criteria = new CoolHitCriteria
            {
                distance = Vector3.Distance(fireEvent.position, hitEvent.position),
                impactVelocity = hitEvent.velocity.magnitude,
                bounceCount = bounceCount,
                predictionScore = CalculatePredictionScore(fireEvent, hitEvent),
                movingTarget = false,  // TODO: Detect if target was moving
                timeToHit = hitEvent.timestamp - fireEvent.timestamp
            };

            bool isCool = IsCoolHit(criteria);

            if (isCool)
            {
                string description = GenerateHitDescription(criteria);
                ExtractClip(hitEvent.timestamp, description, criteria);
            }

            return isCool;
        }

        /// <summary>
        /// Calculate prediction score based on how much the shooter had to lead the target.
        /// </summary>
        private float CalculatePredictionScore(ProjectileEvent fireEvent, ProjectileEvent hitEvent)
        {
            // Simple estimation: distance between fire position and hit position
            // In a full implementation, this would compare against target position at fire time
            Vector3 displacement = hitEvent.position - fireEvent.position;
            return displacement.magnitude;
        }

        /// <summary>
        /// Determine if a hit meets the criteria for a "cool shot".
        /// </summary>
        private bool IsCoolHit(CoolHitCriteria criteria)
        {
            // Long range shot
            if (criteria.distance >= longRangeDistance)
                return true;

            // High velocity impact
            if (criteria.impactVelocity >= highVelocityThreshold)
                return true;

            // Bank shot (multiple bounces)
            if (criteria.bounceCount >= minBouncesForCool)
                return true;

            // Prediction shot (significant lead)
            if (criteria.predictionScore >= predictionThreshold)
                return true;

            // Moving target hit
            if (criteria.movingTarget)
                return true;

            return false;
        }

        /// <summary>
        /// Generate a human-readable description for the cool hit.
        /// </summary>
        private string GenerateHitDescription(CoolHitCriteria criteria)
        {
            List<string> tags = new List<string>();

            if (criteria.distance >= longRangeDistance)
                tags.Add($"Long Range ({criteria.distance:F0}m)");

            if (criteria.impactVelocity >= highVelocityThreshold)
                tags.Add($"High Speed ({criteria.impactVelocity:F0} m/s)");

            if (criteria.bounceCount >= minBouncesForCool)
                tags.Add($"Bank Shot ({criteria.bounceCount} bounces)");

            if (criteria.predictionScore >= predictionThreshold)
                tags.Add("Prediction Shot");

            if (criteria.movingTarget)
                tags.Add("Moving Target");

            if (tags.Count == 0)
                return "Cool Hit";

            return string.Join(" + ", tags);
        }

        /// <summary>
        /// Extract a clip around the cool hit.
        /// </summary>
        private void ExtractClip(float hitTimestamp, string description, CoolHitCriteria criteria)
        {
            if (_clipsThisSession >= maxClipsPerSession)
            {
                Debug.LogWarning("[CoolHitDetector] Max clips per session reached");
                return;
            }

            if (_recorder != null)
            {
                _recorder.ExtractClipAroundEvent(hitTimestamp, clipBeforeHit, clipAfterHit, description);
            }

            DemoClip clip = new DemoClip
            {
                startTime = hitTimestamp - clipBeforeHit,
                endTime = hitTimestamp + clipAfterHit,
                startFrameIndex = 0,  // Will be set by recorder
                endFrameIndex = 0,
                description = description
            };

            _detectedClips.Add(clip);
            _clipsThisSession++;

            OnCoolHitDetected?.Invoke(clip);

            Debug.Log($"[CoolHitDetector] Cool hit detected: {description}");
        }

        /// <summary>
        /// Reset clip counter for new session.
        /// </summary>
        public void ResetSession()
        {
            _clipsThisSession = 0;
            _detectedClips.Clear();
        }

        /// <summary>
        /// Criteria data structure for hit evaluation.
        /// </summary>
        private struct CoolHitCriteria
        {
            public float distance;
            public float impactVelocity;
            public int bounceCount;
            public float predictionScore;
            public bool movingTarget;
            public float timeToHit;
        }
    }
}

File: client/src/demo/management/CoolHitDetector.ts
import * as THREE from 'three';
import { ProjectileEvent } from '../types/ProjectileEvent.js';
import { DemoClip } from '../types/DemoClip.js';

/**
 * Criteria for evaluating if a hit is "cool".
 */
export interface CoolHitCriteria {
  /** Distance from shooter to target */
  distance: number;
  /** Projectile velocity at impact */
  velocity: number;
  /** Number of bounces before hit */
  bounceCount: number;
  /** Whether it was a prediction shot (leading target) */
  isPrediction: boolean;
  /** Whether target was moving */
  targetWasMoving: boolean;
}

/**
 * Configuration for cool hit detection.
 */
export interface CoolHitConfig {
  /** Minimum distance for long range shot (meters) */
  longRangeDistance: number;
  /** Minimum velocity for high velocity shot (m/s) */
  highVelocityThreshold: number;
  /** Minimum bounces for bank shot */
  minBouncesForCool: number;
  /** Whether to enable prediction detection */
  enablePredictionDetection: boolean;
  /** Seconds before hit to include in clip */
  clipBeforeHit: number;
  /** Seconds after hit to include in clip */
  clipAfterHit: number;
  /** Maximum clips per session */
  maxClipsPerSession: number;
}

/**
 * Detects "cool hits" for automatic highlight clip extraction.
 * Evaluates shots based on configurable criteria.
 */
export class CoolHitDetector {
  private config: CoolHitConfig;
  private detectedClips: DemoClip[] = [];
  private projectileBounceCount: Map<number, number> = new Map();

  constructor(config?: Partial<CoolHitConfig>) {
    this.config = {
      longRangeDistance: 100,
      highVelocityThreshold: 50,
      minBouncesForCool: 2,
      enablePredictionDetection: true,
      clipBeforeHit: 2,
      clipAfterHit: 2,
      maxClipsPerSession: 10,
      ...config,
    };
  }

  /**
   * Evaluate a hit event to determine if it's a "cool shot".
   * @param hitEvent - The hit event
   * @param fireEvent - The corresponding fire event
   * @returns Whether the hit is cool
   */
  evaluateHit(
    hitEvent: ProjectileEvent,
    fireEvent: ProjectileEvent
  ): boolean {
    const bounceCount = this.projectileBounceCount.get(hitEvent.projectileId) ?? 0;

    const criteria: CoolHitCriteria = {
      distance: fireEvent.position.distanceTo(hitEvent.position),
      velocity: hitEvent.velocity.length(),
      bounceCount,
      isPrediction: this.calculatePredictionScore(fireEvent, hitEvent) > 0.5,
      targetWasMoving: true, // TODO: Get from target system
    };

    const isCool = this.isCoolHit(criteria);

    if (isCool) {
      const description = this.generateHitDescription(criteria);
      this.extractClip(hitEvent.timestamp, description, criteria);
    }

    // Reset bounce count for this projectile
    this.projectileBounceCount.delete(hitEvent.projectileId);

    return isCool;
  }

  /**
   * Track a bounce event for a projectile.
   * @param projectileId - Projectile ID
   */
  trackBounce(projectileId: number): void {
    const current = this.projectileBounceCount.get(projectileId) ?? 0;
    this.projectileBounceCount.set(projectileId, current + 1);
  }

  /**
   * Calculate prediction score (how much the shot led the target).
   * @param fireEvent - Fire event
   * @param hitEvent - Hit event
   * @returns Prediction score (0-1)
   */
  private calculatePredictionScore(fireEvent: ProjectileEvent, hitEvent: ProjectileEvent): number {
    // Simple implementation: distance-based
    // In a real implementation, this would compare shot direction to target movement
    const distance = fireEvent.position.distanceTo(hitEvent.position);
    const maxDistance = 200; // meters
    return Math.min(1, distance / maxDistance);
  }

  /**
   * Check if criteria meets cool hit thresholds.
   * @param criteria - Hit criteria
   * @returns Whether the hit is cool
   */
  private isCoolHit(criteria: CoolHitCriteria): boolean {
    // Long range shot
    if (criteria.distance >= this.config.longRangeDistance) {
      return true;
    }

    // High velocity hit
    if (criteria.velocity >= this.config.highVelocityThreshold) {
      return true;
    }

    // Bank shot (multiple bounces)
    if (criteria.bounceCount >= this.config.minBouncesForCool) {
      return true;
    }

    // Prediction shot
    if (this.config.enablePredictionDetection && criteria.isPrediction) {
      return true;
    }

    return false;
  }

  /**
   * Generate a description for a cool hit.
   * @param criteria - Hit criteria
   * @returns Description string
   */
  private generateHitDescription(criteria: CoolHitCriteria): string {
    const parts: string[] = [];

    if (criteria.distance >= this.config.longRangeDistance) {
      parts.push('Long Range');
    }

    if (criteria.velocity >= this.config.highVelocityThreshold) {
      parts.push('High Velocity');
    }

    if (criteria.bounceCount >= this.config.minBouncesForCool) {
      parts.push('Bank Shot');
    }

    if (this.config.enablePredictionDetection && criteria.isPrediction) {
      parts.push('Prediction');
    }

    if (criteria.targetWasMoving) {
      parts.push('Moving Target');
    }

    return parts.length > 0 ? parts.join(' + ') : 'Cool Shot';
  }

  /**
   * Extract a clip around a hit event.
   * @param hitTimestamp - Hit event timestamp
   * @param description - Clip description
   * @param criteria - Hit criteria
   */
  private extractClip(hitTimestamp: number, description: string, criteria: CoolHitCriteria): void {
    if (this.detectedClips.length >= this.config.maxClipsPerSession) {
      console.warn('[CoolHitDetector] Max clips reached, skipping extraction');
      return;
    }

    const clip: DemoClip = {
      startTime: hitTimestamp - this.config.clipBeforeHit,
      endTime: hitTimestamp + this.config.clipAfterHit,
      startFrameIndex: 0, // TODO: Calculate from frame buffer
      endFrameIndex: 0, // TODO: Calculate from frame buffer
      description,
    };

    this.detectedClips.push(clip);
    console.log(`[CoolHitDetector] Extracted clip: ${description}`);
  }

  /**
   * Get all detected clips.
   */
  getDetectedClips(): DemoClip[] {
    return [...this.detectedClips];
  }

  /**
   * Reset for a new recording session.
   */
  resetSession(): void {
    this.detectedClips = [];
    this.projectileBounceCount.clear();
  }

  /**
   * Update configuration.
   */
  updateConfig(config: Partial<CoolHitConfig>): void {
    this.config = { ...this.config, ...config };
  }

  /**
   * Get current configuration.
   */
  getConfig(): CoolHitConfig {
    return { ...this.config };
  }
}
